### PR TITLE
SDN controller API javadoc improvements

### PR DESCRIPTION
**Pull request by larkinscarvalho**

### CODE REVIEW

File: src/main/java/org/osc/sdk/controller/FailurePolicyType.java [hunk 1]
```diff
@@ -16,8 +16,24 @@
  *******************************************************************************/
 package org.osc.sdk.controller;
 
+/**
+ * This enum represents the traffic flow action in case of device failure.
+ */
 public enum FailurePolicyType {
-    FAIL_OPEN("FAIL_OPEN"), FAIL_CLOSE("FAIL_CLOSE"), NA("N/A");
+	/**
+	 * Traffic is allowed to pass with no inspection in case of device failure.
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/FailurePolicyType.java]: with no -> without

File: src/main/java/org/osc/sdk/controller/FlowInfo.java [hunk 2]
```diff
@@ -16,15 +16,49 @@
  *******************************************************************************/
 package org.osc.sdk.controller;
 
+/**
+ * This class represents a flow information.
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/FlowInfo.java]: Provide more details here... i.e.: represents a flow information to be used when....

File: src/main/java/org/osc/sdk/controller/FlowPortInfo.java [hunk 3]
```diff
@@ -16,15 +16,29 @@
  *******************************************************************************/
 package org.osc.sdk.controller;
 
+/**
+ * This class represents a flow information with source and destination port identifier.
+ */
 public class FlowPortInfo {
+
+	/**
+	 * Flow information, see {@link FlowInfo}
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/FlowPortInfo.java]: Is this the flow between the source and the destination? If so make that clear.

File: src/main/java/org/osc/sdk/controller/Status.java [hunk 4]
```diff
@@ -16,10 +16,24 @@
  *******************************************************************************/
 package org.osc.sdk.controller;
 
+/**
+ * This class represents the status of SDN controller.
+ */
 public class Status {
 
+	/**
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/Status.java]: wrong identation

File: src/main/java/org/osc/sdk/controller/Status.java [hunk 5]
```diff
@@ -16,10 +16,24 @@
...
+	 */
     private String name;
+
+    /**
+     * The version of SDN controller
+     */
     private String version;
+
+    /**
+     * Is controller ready to provide service
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/Status.java]: Indicates whether the SDN controller is ready to provide service

File: src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java [hunk 6]
```diff
@@ -18,51 +18,54 @@
...
-     * obtained
-     * and the region to operate on
-     *
-     * @param vc
-     *            Virtualization Connector Element
-     * @param region
-     *            region to operate on
+     * @param vc  provides context information of the virtualization connector, see {@link VirtualizationConnectorElement}
+     * @param region  provides information about region to operate on
+     * @return the APIs used by OSC to manage redirection
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java]: redirection -> traffic redirection

File: src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java [hunk 7]
```diff
@@ -18,51 +18,54 @@
...
-     *
-     * @param portsQuery
-     *            Array of flow query info
-     * @return Original input array with source and/or destination ports filled if found. Null if not found.
-     * @throws Exception
-     *             if case of invalid input or service issues (connectivity, etc)
+     * @param vc  provides context information of the virtualization connector, see {@link VirtualizationConnectorElement}
+     * @param region  provides region info to operate on
+     * @param portsQuery  provides a map of context information of the flowInfo, see {@link FlowInfo}
+     * @return a map of FlowPortInfo
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java]: mention what is it mapped by

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 8]
```diff
@@ -33,211 +33,230 @@
 public interface SdnRedirectionApi extends AutoCloseable {
 
     /**
-     * Create NetworkElement with a list of child elements
-     * For example, create NetworkElement with ports of protected workloads
-     * @param elements
-     *          ports of protected workloads or any network element
-     * @return the created NetworkElement
-     * @throws Exception
+     * Creates a NetworkElement
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: Creates a network element
<p>
Describe what is a network element, i.e.: network port, a port group etc. You may need to look at some plugin implementations to come up with a description

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 9]
```diff
@@ -33,211 +33,230 @@
...
      */
     void deleteNetworkElement(NetworkElement element) throws Exception;
 
     /**
-     * Get the children elements of the network element
-     * For example get the protected workload ports of the Port Group
-     * @param element
-     * @return
-     * @throws Exception
+     * Retrieve the list of NetworkElements
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: NetworkElemens -> network elements, see other places as well

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 10]
```diff
@@ -33,211 +33,230 @@
...
+     * @throws NetworkPortNotFoundException when port is not found
+     * @throws Exception upon failure
      */
     String installInspectionHook(NetworkElement networkElem, InspectionPortElement inspectionPort, Long tag,
             TagEncapsulationType encType, Long order, FailurePolicyType failurePolicyType)
                     throws NetworkPortNotFoundException, Exception;
 
     /**
-     * Remove previously installed network hooks placed on behalf of network port.
+     * Deletes an inspection hook
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: Deletes the inspection hook corresponding to the provided network element and inspection port

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 11]
```diff
@@ -33,211 +33,230 @@
...
-     *            The inspection port belonging to network inspection device.
+     * @param networkElem  provides context information of the network element, see {@link NetworkElement}
+     * @param inspectionPort  provides context information of the inspection port element, see{@link InspectionPortElement}
+     * @throws Exception upon failure
      */
     void removeInspectionHook(NetworkElement networkElem, InspectionPortElement inspectionPort) throws Exception;
 
     /**
-     * Remove the previously installed inspection hook given its identifier.
+     * Deleted an inspection hook
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: Deletes the inspection hook given its identifier

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 12]
```diff
@@ -33,211 +33,230 @@
...
      *
-     * @param inspectionHookId
-     *            The identifier of the inspection hook to be removed.
+     * @param inspectionHookId  the identifier of the inspection hook
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: ...hook to be deleted

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 13]
```diff
@@ -33,211 +33,230 @@
...
      */
     void removeInspectionHook(String inspectionHookId) throws Exception;
 
     /**
-     * Retrieves a previously installed inspection hook given its identifier.
+     * Retrieves an InspectionHookElement
      *
-     * @param inspectionHookId
-     *            The identifier of the inspection hook to be retrieved.
+     * @param inspectionHookId  the identifier of the inspection hook
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: ... to be retrieved

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 14]
```diff
@@ -33,211 +33,230 @@
...
+     * @return the inspection hook element, see {@link InspectionHookElement}
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: If there is none found should it return null? Take a look a the OSC core code and see what is the expectation. The javadoc should indicate that null is the return value if nothing is found (if that is what OSC expects)

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 15]
```diff
@@ -33,211 +33,230 @@
...
+     * @throws Exception upon failure
      */
     InspectionHookElement getInspectionHook(String inspectionHookId) throws Exception;
 
     /**
-     * Remove previously installed network hooks placed on behalf of network port.
+     * Removes all the inspection hooks
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: not ALL but all related to the provided network element. Also you should indicate that this should be idempotent (if none is found does nothing and returns successfully, please to the same for the other APIs)

File: src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java [hunk 16]
```diff
@@ -33,211 +33,230 @@
...
-     * Return the current tag used for an installed inspection hook.
+     * Retrieves the inspection hook tag
      *
-     * @param inspectedPort
-     *            The port on which behalf traffic is being inspected.
-     * @param inspectionPort
-     *            The inspection port belonging to network inspection device.
-     * @return The current inspection tag
+     * @param inspectedPort  provides context information of the network element, see {@link NetworkElement}
+     * @param inspectionPort  provides context information of the inspection port element, see{@link InspectionPortElement}
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: I strongly recommend removing the see, look at the resulting javadoc, it already provides the link to the parameter type.

Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnRedirectionApi.java]: Also there are many places where you dont use see... So it is inconsistent.

File: src/main/java/org/osc/sdk/controller/FailurePolicyType.java (hunk 1 above)
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/FailurePolicyType.java]: @larkinscarvalho  missed

File: src/main/java/org/osc/sdk/controller/Status.java (hunk 4 above)
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/Status.java]: missed

File: src/main/java/org/osc/sdk/controller/FailurePolicyType.java [hunk 17]
```diff
@@ -31,7 +31,7 @@
     FAIL_CLOSE("FAIL_CLOSE"),
 
     /**
-     *
+     * N/A in case SDN controller do not support failure policy.
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/FailurePolicyType.java]: do not -> does not

File: src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java [hunk 18]
```diff
@@ -46,25 +46,25 @@
...
      */
     SdnRedirectionApi createRedirectionApi(VirtualizationConnectorElement vc, String region);
 
     /**
-     * @param vc  provides context information of the virtualization connector, see {@link VirtualizationConnectorElement}
+     * @param vc  provides context information of the virtualization connector
      * @param region  provides region info to operate on
-     * @param portsQuery  provides a map of context information of the flowInfo, see {@link FlowInfo}
-     * @return a map of FlowPortInfo
+     * @param portsQuery  provides the context information of the flowInfo mapped by string
```
Comment (emanoelxavier) [inline src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java]: what is supposed to be in the string?

Comment (larkinscarvalho) [inline src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java]: I have no idea. Both NSC and Nuage plugin does not support flow based query.

Comment (larkinscarvalho) [inline src/main/java/org/osc/sdk/controller/api/SdnControllerApi.java]: @arvindn05 can you please provide some inputs here? Thanks

**Pull request merged**